### PR TITLE
MPI: parallel gather/scatter

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

File: devito/dle/backends/basic.py [hunk 1]
```diff
@@ -40,7 +39,7 @@ def _create_efuncs(self, nodes, state):
         """
         noinline = self._compiler_decoration('noinline', c.Comment('noinline?'))
 
-        functions = OrderedDict()
+        efuncs = OrderedDict()
         mapper = {}
         for tree in retrieve_iteration_tree(nodes, mode='superset'):
             # Search an elementizable sub-tree (if any)
```
Comment (georgebisbas) [inline devito/dle/backends/basic.py]: What exactly is meant by the word "elementizable"   ?

File: devito/types/basic.py [hunk 2]
```diff
@@ -713,8 +713,8 @@ class Array(AbstractCachedFunction):
     """
     Tensor symbol representing an array in symbolic equations.
 
-    An Array is created and managed directly by Devito; that is, it is not
-    expected to be needed in user code.
+    Arrays are created and managed directly by Devito (IOW, they are not
+    expected to be used directly in user code).
```
Comment (mloubout) [inline devito/types/basic.py]: Maybe a warning for it?

File: devito/types/basic.py [hunk 3]
```diff
@@ -747,7 +746,7 @@ def __init__(self, *args, **kwargs):
             super(Array, self).__init__(*args, **kwargs)
 
             self._scope = kwargs.get('scope', 'heap')
-            assert self._scope in ['heap', 'stack', 'external']
+            assert self._scope in ['heap', 'stack']
```
Comment (mloubout) [inline devito/types/basic.py]: so external isn't allowed anymore?

File: devito/dle/backends/basic.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/dle/backends/basic.py]: a very bad name I chose long time ago to indicate functions that are called from within the main kernel body

File: devito/types/basic.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/types/basic.py]: to be triggered how/when?

File: devito/types/basic.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/types/basic.py]: correct, it was just redundant -- over time, I got to the point that Arrays are allocated internally by Devito and they never are attached data, so they are implicitly 'external' (see also the inherited ``_mem_external`` property)

**Comment by tjb900** [comment]
OK!  Finally got this going, apologies for the delay.  Seems to work great - makes my demo run much faster (2m04s down from 4m18s, and a lot of that will be fixed overhead before/after the propagation, so I'd say this is at least 3-4x quicker)

I have a couple of quick observations about the generated code, and then I'll go back and look at the actual patches.

```
void gather_u(float *restrict buf_vec, const int buf_x_size, const int buf_y_size, const int buf_z_size, struct dataobj *restrict f_vec, const int otime, const int ox, const int oy, const int oz, const int 
nthreads)
{
  int nt = nthreads;
  float (*restrict f)[f_vec->size[1]][f_vec->size[2]][f_vec->size[3]] __attribute__ ((aligned (64))) = (float (*)[f_vec->size[1]][f_vec->size[2]][f_vec->size[3]]) f_vec->data;
  float (*restrict buf)[buf_y_size][buf_z_size] __attribute__ ((aligned (64))) = (float (*)[buf_y_size][buf_z_size]) buf_vec;
  #pragma omp parallel num_threads(nt)
  {
    #pragma omp for collapse(1) schedule(static)
    for (int x = 0; x <= buf_x_size - 1; x += 1)
    {
      for (int y = 0; y <= buf_y_size - 1; y += 1)
      {
        for (int z = 0; z <= buf_z_size - 1; z += 1)
        {
          buf[x][y][z] = f[otime][x + ox][y + oy][z + oz];
        }
      }
    }
  }
}
```

- `collapse(1)` seems like it could be `collapse(2)`?  I can imagine in some cases the x buffer size will be quite thin, and this will severely limit the parallelism here.

```
void sendrecv_u(struct dataobj *restrict f_vec, const int buf_x_size, const int buf_y_size, const int buf_z_size, const int ogtime, const int ogx, const int ogy, const int ogz, const int ostime, const int o
sx, const int osy, const int osz, const int fromrank, const int torank, MPI_Comm comm, const int nthreads)
{
  float (*bufs)[buf_y_size][buf_z_size];
  float (*bufg)[buf_y_size][buf_z_size];
  posix_memalign((void**)&bufs, 64, sizeof(float[buf_x_size][buf_y_size][buf_z_size]));
  posix_memalign((void**)&bufg, 64, sizeof(float[buf_x_size][buf_y_size][buf_z_size]));
  MPI_Request rrecv;
  MPI_Request rsend;
  MPI_Status srecv;
  MPI_Irecv((float *)bufs,buf_x_size*buf_y_size*buf_z_size,MPI_FLOAT,fromrank,13,comm,&rrecv);
  gather_u((float *)bufg,buf_x_size,buf_y_size,buf_z_size,f_vec,ogtime,ogx,ogy,ogz,nthreads);
  MPI_Isend((float *)bufg,buf_x_size*buf_y_size*buf_z_size,MPI_FLOAT,torank,13,comm,&rsend);
  MPI_Wait(&rsend,MPI_STATUS_IGNORE);
  MPI_Wait(&rrecv,&srecv);
  if (fromrank != MPI_PROC_NULL)
  {
    scatter_u((float *)bufs,buf_x_size,buf_y_size,buf_z_size,f_vec,ostime,osx,osy,osz,nthreads);
  }
  free(bufs);
  free(bufg);
}
```

* I'm happy to leave it for another PR - it's just a performance optimisation - but I think that the `gather_u` should only be done if `torank != MPI_PROC_NULL`.  This represents a lot of wasted work for small `mpi_size`s where most of the faces are not exchanged.
* Nit: why do you collect the status of the receive but not the send?

**Comment by FabioLuporini** [comment]
Merged.

**Comment by FabioLuporini** [comment]
Thanks for the insights @tjb900 . Since I'm working on the same kind of stuff in another branch, and since what you pointed out is a non-stopper, I'll deal with it in my next PR. Thanks again !

**Pull request merged**